### PR TITLE
[native-bridge] - committee registration cli tool

### PR DESCRIPTION
**Pull request by patrickkuo**
## Description 

added a new command to bridge tool cli to register as a committee member

## Test plan 

manually tested on bridgenet

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

**Review by dariorussi — APPROVED** [review]
this looks fine to me with the flags being a bit verbose :)
It would be good for @longbowlu to have a look at this too.
The idea here is that a validator would run this command to register themselves?
Can you explain the flow? (sorry I am not sure what that is)

File: crates/sui-bridge/src/tools/cli.rs [hunk 1]
```diff
@@ -155,6 +167,85 @@ async fn main() -> anyhow::Result<()> {
 
             return Ok(());
         }
+        BridgeValidatorCommand::CommitteeRegistration {
```
Comment (longbowlu) [inline crates/sui-bridge/src/tools/cli.rs:170]: I'm not sure this file is the best place to add this command:
1. IMO we shouldn't ask for a `NodeConfig` which is a huge file but the only needed field is authority account key.
2. This file reminds me of `validator_commands.rs`: https://github.com/MystenLabs/sui/blob/main/crates/sui/src/validator_commands.rs#L221. This files contains some cli tools which validators can use for daily operations. The way it works is it loads rpc url and key as `WalletContext`. 


thoughts?

**Review by longbowlu — CHANGES_REQUESTED** [review]

File: crates/sui-bridge/src/tools/cli.rs (hunk 1 above)
Comment (patrickkuo) [reply crates/sui-bridge/src/tools/cli.rs:170]: ya this seem more suitable in validator commands, I can move it there.

the reason I load the node config is in our bridgenet setup, validator's account key can only be found in this file, and it's not in client.yml, I will see if we can change the deployment to add the account key to client.yaml.

File: crates/sui/src/validator_commands.rs [hunk 2]
```diff
@@ -163,6 +169,12 @@ pub enum SuiValidatorCommand {
         #[clap(name = "gas-budget", long)]
         gas_budget: Option<u64>,
     },
+    BridgeCommitteeRegistration {
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:172]: Doc comments for clap please

File: crates/sui/src/validator_commands.rs [hunk 3]
```diff
@@ -163,6 +169,12 @@ pub enum SuiValidatorCommand {
...
+        #[clap(long = "bridge-node-config-path")]
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:173]:
```suggestion
        #[clap(long)]
```

File: crates/sui/src/validator_commands.rs [hunk 4]
```diff
@@ -163,6 +169,12 @@ pub enum SuiValidatorCommand {
...
+        bridge_node_config_path: PathBuf,
+        #[clap(long = "bridge-authority-url")]
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:175]:
```suggestion
        #[clap(long)]
```

File: crates/sui/src/validator_commands.rs [hunk 5]
```diff
@@ -454,6 +467,66 @@ impl SuiValidatorCommand {
...
+                bridge_authority_url,
+            } => {
+                let bridge_config = BridgeNodeConfig::load(bridge_node_config_path)
+                    .expect("Couldn't load BridgeNodeConfig");
+
+                // Read bridge keypair
+                let key_str =
+                    fs::read_to_string(bridge_config.bridge_authority_key_path_base64_raw)?;
+                let bytes = Base64::decode(key_str.trim()).unwrap();
+                let key = Secp256k1PrivateKey::from_bytes(&bytes).unwrap();
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:481]: bail with message on error!

File: crates/sui/src/validator_commands.rs [hunk 6]
```diff
@@ -454,6 +467,66 @@ impl SuiValidatorCommand {
...
+                    Identifier::new("committee_registration").unwrap(),
+                    vec![],
+                    vec![bridge_arg, system_arg, pub_key_arg, url_arg],
+                );
+
+                let tx_data = TransactionData::new_programmable(
+                    address,
+                    vec![gas],
+                    ptb.finish(),
+                    10000000,
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:518]: Will this gas be always enough?

File: crates/sui/src/validator_commands.rs [hunk 7]
```diff
@@ -454,6 +467,66 @@ impl SuiValidatorCommand {
...
+                    sui_bridge::sui_client::SuiClient::new(&bridge_config.sui.sui_rpc_url).await?;
+                let bridge = bridge_client
+                    .get_mutable_bridge_object_arg_must_succeed()
+                    .await;
+
+                let gas = context
+                    .get_one_gas_object_owned_by_address(address)
+                    .await?
+                    .unwrap();
+                let ref_gas_price = context.get_reference_gas_price().await?;
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:498]: nit: we use `gas_price` everywhere in this codebase, so I'd suggest we keep it consistent.
```suggestion
                let gas_price = context.get_reference_gas_price().await?;
```

File: crates/sui/src/validator_commands.rs [hunk 8]
```diff
@@ -454,6 +467,66 @@ impl SuiValidatorCommand {
...
+                    vec![gas],
+                    ptb.finish(),
+                    10000000,
+                    ref_gas_price,
+                );
+
+                let tx = context.sign_transaction(&tx_data);
+                let response = context.execute_transaction_must_succeed(tx).await;
+                println!(
+                    "Committee registration successful. txn digest: {}",
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:525]:
```suggestion
                    "Committee registration successful. Transaction digest: {}",
```

File: crates/sui/src/validator_commands.rs [hunk 9]
```diff
@@ -689,6 +762,9 @@ impl Display for SuiValidatorCommandResponse {
                     data, serialized_data
                 )?;
             }
+            SuiValidatorCommandResponse::BridgeCommitteeRegistration(response) => {
+                write!(writer, "{}", write_transaction_response(response)?)?;
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:766]: nit: I would suggest to just use the standard Display for SuiTransactionBlockResponse, so just do. However, I see that validator's commands output has not been migrated to table format, so probably this is fine for now.
```suggestion
                write!(writer, "{}", response)?;
```

File: crates/sui/src/validator_commands.rs [hunk 10]
```diff
@@ -454,6 +467,66 @@ impl SuiValidatorCommand {
...
+                let bytes = Base64::decode(key_str.trim()).unwrap();
+                let key = Secp256k1PrivateKey::from_bytes(&bytes).unwrap();
+                let ecdsa_keypair = Secp256k1KeyPair::from(key);
+
+                // Read sui node account key pair
+                let address = context.active_address()?;
+                println!("Starting bridge committee registration for Sui validator: {address}, with bridge public key: {}", ecdsa_keypair.public);
+
+                let bridge_client =
+                    sui_bridge::sui_client::SuiClient::new(&bridge_config.sui.sui_rpc_url).await?;
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:489]: nit: import `sui_bridge::sui_client::SuiClient as BridgeSuiClient`

File: crates/sui/src/validator_commands.rs [hunk 11]
```diff
@@ -454,6 +467,66 @@ impl SuiValidatorCommand {
...
+                let bridge = bridge_client
+                    .get_mutable_bridge_object_arg_must_succeed()
+                    .await;
+
+                let gas = context
+                    .get_one_gas_object_owned_by_address(address)
+                    .await?
+                    .unwrap();
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:497]: Can we handle this unwrap accordingly and bail on error?

File: crates/sui/src/validator_commands.rs [hunk 12]
```diff
@@ -454,6 +467,66 @@ impl SuiValidatorCommand {
...
+                    .expect("Couldn't load BridgeNodeConfig");
+
+                // Read bridge keypair
+                let key_str =
+                    fs::read_to_string(bridge_config.bridge_authority_key_path_base64_raw)?;
+                let bytes = Base64::decode(key_str.trim()).unwrap();
+                let key = Secp256k1PrivateKey::from_bytes(&bytes).unwrap();
+                let ecdsa_keypair = Secp256k1KeyPair::from(key);
+
+                // Read sui node account key pair
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:484]: This comment does not seem right.

File: crates/sui/src/validator_commands.rs [hunk 13]
```diff
@@ -454,6 +467,66 @@ impl SuiValidatorCommand {
...
+                let mut ptb = ProgrammableTransactionBuilder::default();
+
+                let bridge_arg = ptb.obj(bridge)?;
+                let system_arg = ptb.obj(ObjectArg::SUI_SYSTEM_MUT)?;
+                let pub_key_arg = ptb.pure(ecdsa_keypair.public().as_bytes())?;
+                let url_arg = ptb.pure(bridge_authority_url)?;
+                ptb.programmable_move_call(
+                    BRIDGE_PACKAGE_ID,
+                    BRIDGE_MODULE_NAME.into(),
+                    Identifier::new("committee_registration").unwrap(),
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:509]: Handle unwrap and bail on error.

File: crates/sui/src/validator_commands.rs [hunk 14]
```diff
@@ -454,6 +467,66 @@ impl SuiValidatorCommand {
                     serialized_data,
                 }
             }
+            SuiValidatorCommand::BridgeCommitteeRegistration {
+                bridge_node_config_path,
+                bridge_authority_url,
+            } => {
+                let bridge_config = BridgeNodeConfig::load(bridge_node_config_path)
+                    .expect("Couldn't load BridgeNodeConfig");
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:475]: Can we add more explanation to this error. Is the path wrong, is the path missing, etc? What should the user do when this error happens?

File: crates/sui/src/validator_commands.rs [hunk 15]
```diff
@@ -454,6 +467,66 @@ impl SuiValidatorCommand {
...
+
+                // Read bridge keypair
+                let key_str =
+                    fs::read_to_string(bridge_config.bridge_authority_key_path_base64_raw)?;
+                let bytes = Base64::decode(key_str.trim()).unwrap();
```
Comment (stefan-mysten) [inline crates/sui/src/validator_commands.rs:480]: nit: handle unwrap better and bail when error?

**Review by stefan-mysten — COMMENTED** [review]
This is looks pretty good but left a few comments with regards to proper error handling. Please have a look!

File: crates/sui/src/validator_commands.rs (hunk 6 above)
Comment (patrickkuo) [reply crates/sui/src/validator_commands.rs:518]: will make this configurable

File: crates/sui/src/validator_commands.rs [hunk 16]
```diff
@@ -454,6 +474,70 @@ impl SuiValidatorCommand {
...
+                    .await;
+
+                let gas = context
+                    .get_one_gas_object_owned_by_address(address)
+                    .await?
+                    .unwrap_or_else(|| panic!("Cannot find gas object from address : {address}"));
+
+                let gas_price = context.get_reference_gas_price().await?;
+
+                let mut ptb = ProgrammableTransactionBuilder::default();
```
Comment (longbowlu) [inline crates/sui/src/validator_commands.rs:510]: you can call this function https://github.com/MystenLabs/sui/blob/5db2bb5d86dd4e8e48678bfe08272fc5d37379a6/crates/sui-bridge/src/sui_transaction_builder.rs#L550

File: crates/sui/src/validator_commands.rs [hunk 17]
```diff
@@ -454,6 +474,70 @@ impl SuiValidatorCommand {
...
+                    Err(e) => panic!("Couldn't load BridgeNodeConfig, caused by: {e}"),
+                };
+                // Read bridge keypair
+                let key_str =
+                    fs::read_to_string(bridge_config.bridge_authority_key_path_base64_raw)?;
+                let bytes = Base64::decode(key_str.trim())
+                    .expect("Decode authority key from file should not fail.");
+                let key = Secp256k1PrivateKey::from_bytes(&bytes)
+                    .expect("Decode authority key from bytes should not fail.");
+                let ecdsa_keypair = Secp256k1KeyPair::from(key);
```
Comment (longbowlu) [inline crates/sui/src/validator_commands.rs:493]: use https://github.com/MystenLabs/sui/blob/5db2bb5d86dd4e8e48678bfe08272fc5d37379a6/crates/sui-bridge/src/config.rs#L392

Comment (longbowlu) [reply crates/sui/src/validator_commands.rs:487]: @patrickkuo  one more comment here in case you missed it

**Review by longbowlu — APPROVED** [review]
thanks1

**Pull request merged**